README.md

# saeHB.unit

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/saeHB.unit)](https://CRAN.R-project.org/package=saeHB.unit)
[![check-standard](https://github.com/Alfrzlp/saeHB.unit/actions/workflows/check-standard.yaml/badge.svg)](https://github.com/Alfrzlp/saeHB.unit/actions/workflows/check-standard.yaml)

<!-- badges: end -->

# Author

Azka Ubaidillah, Ridson Al Farizal P

# Maintainer

Ridson Al Farizal P \<alfrzlp@gmail.com\>

# Description

We designed this package to provide several functions for unit level of
small area estimation (Battese, Harter and Fuller model) using
hierarchical Bayesian (HB) method. This package also provides a dataset
produced by a data generation. The `rjags` package is employed to obtain
parameter estimates

# Installation

You can install the development version of saeHB.unit from
[GitHub](https://github.com/) with:

``` r
install.packages("devtools")
devtools::install_github("Alfrzlp/saeHB.unit")
```

Or you can install cran version with

``` r
install.packages(saeHB.unit)
```

## Example 1

This is a basic example which shows you how to solve a common problem:

``` r
library(dplyr)
library(saeHB.unit)
library(ggplot2)
library(tidyr)

windowsFonts(
  poppins = windowsFont('poppins'),
  tnr = windowsFont('Times New Roman')
)
```

### Data

``` r
glimpse(cornsoybean)
#> Rows: 37
#> Columns: 5
#> $ County      <int> 1, 2, 3, 4, 4, 5, 5, 5, 6, 6, 6, 7, 7, 7, 8, 8, 8, 9, 9, 9…
#> $ CornHec     <dbl> 165.76, 96.32, 76.08, 185.35, 116.43, 162.08, 152.04, 161.…
#> $ SoyBeansHec <dbl> 8.09, 106.03, 103.60, 6.47, 63.82, 43.50, 71.43, 42.49, 10…
#> $ CornPix     <int> 374, 209, 253, 432, 367, 361, 288, 369, 206, 316, 145, 355…
#> $ SoyBeansPix <int> 55, 218, 250, 96, 178, 137, 206, 165, 218, 221, 338, 128, …
glimpse(cornsoybeanmeans)
#> Rows: 12
#> Columns: 6
#> $ CountyIndex           <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12
#> $ CountyName            <fct> CerroGordo, Hamilton, Worth, Humboldt, Franklin,…
#> $ SampSegments          <int> 1, 1, 1, 2, 3, 3, 3, 3, 4, 5, 5, 6
#> $ PopnSegments          <dbl> 545, 566, 394, 424, 564, 570, 402, 567, 687, 569…
#> $ MeanCornPixPerSeg     <dbl> 295.29, 300.40, 289.60, 290.74, 318.21, 257.17, …
#> $ MeanSoyBeansPixPerSeg <dbl> 189.70, 196.65, 205.28, 220.22, 188.06, 247.13, …
```

``` r
Xarea <- cornsoybeanmeans %>%
   dplyr::select(
      County = CountyIndex,
      CornPix = MeanCornPixPerSeg,
      SoyBeansPix = MeanSoyBeansPixPerSeg
   )
head(Xarea)
#>   County CornPix SoyBeansPix
#> 1      1  295.29      189.70
#> 2      2  300.40      196.65
#> 3      3  289.60      205.28
#> 4      4  290.74      220.22
#> 5      5  318.21      188.06
#> 6      6  257.17      247.13
```

### Model

``` r
corn_model <- hb_BHF(
   CornHec ~ SoyBeansPix + CornPix,
   data_unit = cornsoybean,
   data_area = Xarea,
   domain = "County",
   iter.update = 20
)
```

![](man/figures/README-unnamed-chunk-7-1.png)<!-- -->![](man/figures/README-unnamed-chunk-7-2.png)<!-- -->

    #>                   Mean         SD       2.5%        25%        50%        75%
    #> intercept    0.1552265  0.7415844 -1.3078113 -0.3262636  0.1458748  0.6601395
    #> SoyBeansPix  0.0045318  0.0044236 -0.0041645  0.0016640  0.0045969  0.0074832
    #> CornPix      0.4002724  0.0025485  0.3951512  0.3985503  0.4002998  0.4019972
    #>              97.5%
    #> intercept   1.6317
    #> SoyBeansPix 0.0132
    #> CornPix     0.4053

## Example 2

### Data

``` r
head(dummy_unit)
#>      domain     y_di       x1       x2
#> 181      d1 72.67585 13.88978 14.35716
#> 279      d1 63.10395 10.38460 13.58800
#> 343      d1 80.10076 16.36110 14.59864
#> 539      d1 68.83004 13.33046 14.17516
#> 670      d1 77.48515 18.62796 12.55454
#> 1424     d1 78.00826 13.44454 16.34891
```

``` r
head(dummy_area)
#>   domain       x1       x2 parameter
#> 1     d1 15.05076 14.96766  77.03467
#> 2     d2 15.07153 14.98934  74.67858
#> 3     d3 14.96426 14.94145  73.35885
#> 4     d4 15.03803 15.02529  77.99655
#> 5     d5 14.98165 14.99815  76.76959
#> 6     d6 15.04244 15.00129  77.30116
```

### Model

``` r
hb_model <- hb_BHF(
  formula = y_di ~ x1 + x2,
  data_unit = dummy_unit,
  data_area = dummy_area,
  domain = "domain",
  iter.update = 30,
  plot = FALSE
)
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 2/30 | ■■■ 7% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 3/30 | ■■■■ 10% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 4/30 | ■■■■■ 13% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 5/30 | ■■■■■■ 17% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 6/30 | ■■■■■■■ 20% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 7/30 | ■■■■■■■■ 23% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 8/30 | ■■■■■■■■■ 27% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 9/30 | ■■■■■■■■■■ 30% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 10/30 | ■■■■■■■■■■■ 33% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 11/30 | ■■■■■■■■■■■■ 37% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 12/30 | ■■■■■■■■■■■■■ 40% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 13/30 | ■■■■■■■■■■■■■■ 43% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 14/30 | ■■■■■■■■■■■■■■■ 47% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 15/30 | ■■■■■■■■■■■■■■■■ 50% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 16/30 | ■■■■■■■■■■■■■■■■■ 53% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 17/30 | ■■■■■■■■■■■■■■■■■■ 57% | ETA: 1m
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 18/30 | ■■■■■■■■■■■■■■■■■■■ 60% | ETA: 50s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 19/30 | ■■■■■■■■■■■■■■■■■■■■ 63% | ETA: 46s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 20/30 | ■■■■■■■■■■■■■■■■■■■■■ 67% | ETA: 42s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 21/30 | ■■■■■■■■■■■■■■■■■■■■■■ 70% | ETA: 38s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 22/30 | ■■■■■■■■■■■■■■■■■■■■■■■ 73% | ETA: 34s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 23/30 | ■■■■■■■■■■■■■■■■■■■■■■■■ 77% | ETA: 30s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 24/30 | ■■■■■■■■■■■■■■■■■■■■■■■■■ 80% | ETA: 25s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 25/30 | ■■■■■■■■■■■■■■■■■■■■■■■■■■ 83% | ETA:
#> 21s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 26/30 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■ 87% | ETA:
#> 17s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 27/30 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 90% |
#> ETA: 13s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 28/30 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 93% |
#> ETA: 8s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> Update 29/30 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 97%
#> | ETA: 4s
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> 
#> Warning in FUN(X[[i]], ...): Failed to set trace monitor for muT
#> Variable muT not found
#> 
#> ── Coefficient ─────────────────────────────────────────────────────────────────
#>                Mean        SD      2.5%       25%       50%       75%  97.5%
#> intercept 0.3888520 0.0603124 0.2736811 0.3463387 0.3884280 0.4297975 0.5090
#> x1        2.0298767 0.0023656 2.0253100 2.0283008 2.0298457 2.0314584 2.0346
#> x2        3.0232782 0.0021441 3.0189985 3.0218045 3.0232850 3.0248043 3.0273
```

## Autocorelation, Trace and Density plot

``` r
saeHB.unit::autoplot(hb_model)
```

![](man/figures/README-unnamed-chunk-11-1.png)<!-- -->![](man/figures/README-unnamed-chunk-11-2.png)<!-- -->
\## Coefficients

``` r
summary(hb_model)
#>                Mean        SD      2.5%       25%       50%       75%  97.5%
#> intercept 0.3888520 0.0603124 0.2736811 0.3463387 0.3884280 0.4297975 0.5090
#> x1        2.0298767 0.0023656 2.0253100 2.0283008 2.0298457 2.0314584 2.0346
#> x2        3.0232782 0.0021441 3.0189985 3.0218045 3.0232850 3.0248043 3.0273
```

``` r
data.frame(
  id = 1:30,
  hb = hb_model$Est$MEAN,
  parameter = dummy_area$parameter
) %>%
  pivot_longer(-1, names_to = "metode", values_to = "rse") %>%
  ggplot(aes(x = id, y = rse, col = metode)) +
  geom_line() +
  scale_color_discrete(
    labels = c('HB with NA', 'Parameter')
  ) +
  labs(col = NULL, y = 'Estimate', x = 'County', title = 'Comparison of estimates') +
  theme(
    text = element_text(family = 'poppins'),
    axis.ticks.x = element_blank(),
    plot.title = element_text(face = 2, vjust = 0),
    plot.subtitle = element_text(colour = 'gray30', vjust = 0)
  )
```

![](man/figures/README-unnamed-chunk-13-1.png)<!-- -->

# References

- Battese, G. E., Harter, R. M., & Fuller, W. A. (1988). An
  error-components model for prediction of county crop areas using
  survey and satellite data. Journal of the American Statistical
  Association, 83(401), 28-36.

- Rao, J. N., & Molina, I. (2015). Small area estimation. John Wiley &
  Sons.


NEWS.md
# saeHB.unit 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/data.R
#' cornsoybean
#' @name cornsoybean
#' @title Corn and soy beans survey and satellite data in 12 counties in Iowa
#' @description Survey and satellite data for corn and soy beans in 12 Iowa counties, obtained from the 1978 June Enumerative Survey of the U.S. Department of Agriculture and from land observatory satellites (LANDSAT) during the 1978 growing season.
#' @format A data frame with 37 observations on the following 5 variables.
#' \describe{
#'    \item{\code{County}:}{ numeric county code.}
#'    \item{\code{CornHec}:}{ reported hectares of corn from the survey.}
#'    \item{\code{SoyBeansHec}:}{ reported hectares of soy beans from the survey.}
#'    \item{\code{CornPix}:}{ number of pixels of corn in sample segment within county, from satellite data. }
#'    \item{\code{SoyBeansPix}:}{ number of pixels of soy beans in sample segment within county, from satellite data.}
#' }
#' @source Battesse, G.E., Harter, R.M. and Fuller, W.A. (1988). An Error-Components Model for Prediction of County Crop Areas Using Survey and Satellite Data. Journal of the American Statistical Association 83, 28-36.
"cornsoybean"

#' cornsoybeanmeans
#' @name cornsoybeanmeans
#' @title Corn and soy beans mean number of pixels per segment for 12 counties in Iowa.
#' @description County means of number of pixels per segment of corn and soy beans, from satellite data, for 12 counties in Iowa. Population size, sample size and means of auxiliary variables in data set \code{\link{cornsoybean}}.
#' @format
#' A data frame with 12 observations on the following 6 variables.
#' \describe{
#'      \item{\code{CountyIndex}:}{ numeric county code.}
#'      \item{\code{CountyName}:}{ name of the county.}
#'      \item{\code{SampSegments}:}{ number of sample segments in the county (sample size).}
#'      \item{\code{PopnSegments}:}{ number of population segments in the county (population size).}
#'      \item{\code{MeanCornPixPerSeg}:}{ mean number of corn pixels per segment in the county.}
#'      \item{\code{MeanSoyBeansPixPerSeg}:}{ mean number of soy beans pixels per segment in the county.}
#'  }
#' @source Battesse, G.E., Harter, R.M. and Fuller, W.A. (1988). An Error-Components Model for Prediction of County Crop Areas Using Survey and Satellite Data. Journal of the American Statistical Association 83, 28-36.
"cornsoybeanmeans"

#' dummy_unit
#' @name dummy_unit
#' @title dummy_unit
#' @description dummy data
#' @format A data frame with 1000 observations on the following 4 variables.
#' \describe{
#'    \item{\code{domain}:}{domain code}
#'    \item{\code{y_di}:}{direct estimate of y}
#'    \item{\code{x1}:}{x1}
#'    \item{\code{x2}:}{x2}
#' }
"dummy_unit"

#' dummy_area
#' @name dummy_area
#' @title dummy_area
#' @description dummy data
#' @format A data frame with 30 observations on the following 4 variables.
#' \describe{
#'    \item{\code{domain}:}{domain code}
#'    \item{\code{x1}:}{x1}
#'    \item{\code{x2}:}{x2}
#'    \item{\code{parameter}:}{ true value of y }
#' }
"dummy_area"


R/autoplot.R
#' Autoplot
#'
#' @param object HB model
#' @param ... other argument
#' @return plot
#'
#' @examples
#' library(dplyr)
#'
#' Xarea <- cornsoybeanmeans %>%
#'    dplyr::select(
#'       County = CountyIndex,
#'       CornPix = MeanCornPixPerSeg,
#'       SoyBeansPix = MeanSoyBeansPixPerSeg
#'    )
#'
#' corn_model <- hb_BHF(
#'    CornHec ~ SoyBeansPix + CornPix,
#'    data_unit = cornsoybean,
#'    data_area = Xarea,
#'    domain = "County",
#'    iter.update = 20,
#'    plot = FALSE
#' )
#' autoplot(corn_model)
#'
#' @export
autoplot.saehb <- function(object, ...){
  graphics::par(mar = c(2, 2, 2, 2))
  coda::autocorr.plot(object$result_mcmc, col = "brown2", lwd = 2)
  plot(object$result_mcmc, col = "brown2", lwd = 2)
}

#' @inherit ggplot2::autoplot
#' @export
autoplot <- function(object, ...) {
  UseMethod("autoplot")
}


R/hb_BHF.R
#' Basic Unit Level Model (Battese-Harter-Fuller model) using Hierarchical Bayesian Approach
#'
#' @description This function gives the Hierarchical Bayesian (HB) based on a basic unit level model (Battese-Harter-Fuller model).
#'
#' @references
#' \enumerate{
#'  \item Battese, G. E., Harter, R. M., & Fuller, W. A. (1988). An error-components model for prediction of county crop areas using survey and satellite data. Journal of the American Statistical Association, 83(401), 28-36.
#'  \item Rao, J. N., & Molina, I. (2015). Small area estimation. John Wiley & Sons.
#' }
#'
#' @param formula an object of class formula that contains a description of the model to be fitted. The variables included in the formula must be contained in the data.
#' @param data_unit data frame containing the variables named in \code{formula} and \code{domain}.
#' @param data_area data frame containing the variables named in \code{formula} and \code{domain}. Each remaining column contains the population means of each of the p auxiliary variables for the D domains.
#' @param domain Character or formula for domain column names in unit data \code{data_unit} and area data \code{data_area}. (example : "County" or ~County)
#' @param iter.update Number of updates with default 3
#' @param iter.mcmc Number of total iterations per chain with default 10000
#' @param coef a vector contains prior initial value of Coefficient of Regression Model for fixed effect with default vector of 0 with the length of the number of regression coefficients
#' @param var.coef a vector contains prior initial value of variance of Coefficient of Regression Model with default vector of 1 with the length of the number of regression coefficients
#' @param thin Thinning rate, must be a positive integer with default 2
#' @param burn.in Number of iterations to discard at the beginning with default 2000
#' @param tau.u Prior initial value of inverse of Variance of area random effect with default 1
#' @param seed number used to initialize a pseudorandom number generator (default seed = 1). The random number generator method used is "base::Wichmann-Hill".
#' @param quiet if TRUE, then messages generated during compilation will be suppressed (default TRUE).
#' @param plot if TRUE, the autocorrelation, trace, and density plots will be generated (default TRUE).
#'
#' @return The function returns a list with the following objects : Estimation \code{Est}, random effect variance \code{refVar}, beta coefficient \code{Coefficient} and MCMC result \code{result_mcmc}
#'
#' @export
#'
#' @examples
#' library(dplyr)
#'
#' Xarea <- cornsoybeanmeans %>%
#'    dplyr::select(
#'       County = CountyIndex,
#'       CornPix = MeanCornPixPerSeg,
#'       SoyBeansPix = MeanSoyBeansPixPerSeg
#'    )
#'
#' corn_model <- hb_BHF(
#'    CornHec ~ SoyBeansPix + CornPix,
#'    data_unit = cornsoybean,
#'    data_area = Xarea,
#'    domain = "County",
#'    iter.update = 20
#' )
#'

hb_BHF <- function(formula, data_unit, data_area, domain, iter.update = 3, iter.mcmc = 10000, coef, var.coef, thin = 3, burn.in = 2000, tau.u = 1, seed = 1, quiet = TRUE, plot = TRUE){
  result <- list(Est = NA, refVar = NA, coefficient = NA, result_mcmc = NA)
  formuladata <- stats::model.frame(formula, data_unit, na.action = NULL)

  # Pengecekan input ---------------------
  if (any(is.na(formuladata[, -1]))) {
    stop("Auxiliary Variables contains NA values.")
  }
  # banyaknya variabel (termasuk Beta0)
  nvar <- ncol(formuladata)

  if (!missing(var.coef)) {
    if (length(var.coef) != nvar) {
      stop("length of vector var.coef does not match the number of regression coefficients, the length must be ", nvar)
    }
    tau.b <- 1 / var.coef
  } else {
    tau.b <- 1 / rep(1, nvar)
  }

  if (!missing(coef)) {
    if (length(coef) != nvar) {
      stop("length of vector coef does not match the number of regression coefficients, the length must be ", nvar)
    }
    mu.b <- coef
  } else {
    mu.b <- rep(0, nvar)
  }
  if (iter.update < 3) {
    stop("the number of iteration updates at least 3 times")
  }

  if (quiet) {
    cli::cli_progress_bar(
      total = iter.update,
      format = "Update {iter}/{iter.update} | {cli::pb_bar} {cli::pb_percent} | ETA: {cli::pb_eta} \n"
    )
    my_env <- environment()
    default_pb <- "none"
    update_progress <- function(iter, iter.update){
      Sys.sleep(2/10000)
      cli::cli_progress_update(set = iter, .envir = my_env)
    }
  }else{
    default_pb <- "text"
    update_progress <- function(iter, iter.update){
      cli::cli_h1('Update {iter}/{iter.update}')
    }
  }


  # Model ------------------
  # Jika tidak ada NA
  if (!any(is.na(formuladata[, 1]))) {
    formuladata <- as.matrix(stats::na.omit(formuladata))

    J <- nrow(data_unit)
    d <- .get_variable(data_unit, domain)
    m <- nrow(data_area)

    X_unit <- stats::model.matrix(formula, data = as.data.frame(formuladata))
    X_unit <- as.matrix(X_unit)
    X_area <- stats::model.matrix(stats::update(formula, NULL ~ .), data = as.data.frame(data_area))
    X_area <- as.matrix(X_area)

    tau.ua <- tau.ub <- tau.ea <- tau.eb <- 1
    a.var <- 1

    my_model <- "model {
				for (j in 1:J) {
						y[j] ~ dnorm(Mu[j],tau.e)
						Mu[j] <- b[1] + sum(b[2:nvar] * x[j,]) + u[d[j]]
				}

				for(i in 1:m){
					u[i] ~ dnorm(0,tau.u)
					mu[i] <- b[1] + sum(b[2:nvar] * X[i,]) + u[i]
				}

				for (k in 1:nvar){
					 b[k] ~ dnorm(mu.b[k], tau.b[k])
				}

				tau.e ~ dgamma(tau.ea,tau.eb)
				tau.u ~ dgamma(tau.ua,tau.ub)
				a.var <- 1 / tau.u
		}"


    for (iter in 1:iter.update) {
      update_progress(iter, iter.update)

      dat <- list(
        nvar = nvar,
        J = J, m = m, d = d,
        y = formuladata[, 1],
        x = as.matrix(X_unit[, -1]),
        X = as.matrix(X_area[, -1]),
        mu.b = mu.b,
        tau.b = tau.b,
        tau.ua = tau.ua, tau.ub = tau.ub,
        tau.ea = tau.ea, tau.eb = tau.eb
      )

      inits <- list(
        u = rep(0, m),
        b = mu.b,
        tau.u = 1,
        tau.e = 1,
        .RNG.name = "base::Wichmann-Hill",
        .RNG.seed = seed
      )

      jags.m <- rjags::jags.model(
        file = textConnection(my_model),
        data = dat, inits = inits,
        n.chains = 1, n.adapt = 500,
        quiet = quiet
      )

      params <- c("mu", "a.var", "b", "tau.u", "tau.e")
      samps1 <- rjags::coda.samples(
        jags.m, params, n.iter = iter.mcmc,
        thin = thin, progress.bar = default_pb
      )
      samps11 <- stats::window(samps1, start = burn.in + 1, end = iter.mcmc - 1)
      hasil <- summary(samps11)

      a.var <- hasil$statistics[1]
      beta <- hasil$statistics[2:(nvar + 1), 1:2]
      mu.b <- beta[, 1]
      tau.b <- 1 / (beta[, 2]^2)

      tau.e <- hasil$statistics[nvar + m + 2, 1:2]
      tau.u <- hasil$statistics[nvar + m + 3, 1:2]

      tau.ea <- tau.e[1]^2 / tau.e[2]^2
      tau.eb <- tau.e[1] / tau.e[2]^2
      tau.ua <- tau.u[1]^2 / tau.u[2]^2
      tau.ub <- tau.u[1] / tau.u[2]^2
    }

    result_samps <- summary(samps1)
    b.varnames <- c('intercept', all.vars(formula[[3]]))

    result_mcmc <- samps1[, c(2:(nvar + 1))]
    colnames(result_mcmc[[1]]) <- b.varnames
    a.var <- result_samps$statistics[1]
    beta <- result_samps$statistics[2:(nvar + 1), 1:2]
    rownames(beta) <- b.varnames
    mu <- result_samps$statistics[(nvar + 2):(1 + nvar + m), 1:2]

    Estimation <- data.frame(mu)
    Quantiles <- as.data.frame(result_samps$quantiles[2:(1 + nvar + m), ])
    q_mu <- Quantiles[-c(1:nvar), ]
    q_beta <- Quantiles[1:nvar, ]
    rownames(q_beta) <- b.varnames
    beta <- cbind(beta, q_beta)
    Estimation <- data.frame(Estimation, q_mu)
    colnames(Estimation) <- c(
      "MEAN", "SD", "2.5%", "25%",
      "50%", "75%", "97.5%"
    )
    rownames(Estimation) <- .get_variable(data_area, domain)
  }else{
    # Jika ada NA
    J <- nrow(data_unit)
    m <- nrow(data_area)

    unit_sampled <- stats::na.omit(data_unit)
    M1 <- nrow(unit_sampled)
    d1 <- .get_variable(unit_sampled, domain)

    data_area$idx <- 1:m
    area_sampled <- stats::na.omit(data_area)
    # area_ns <- dplyr::filter(data_area, !idx %in% area_sampled$idx)
    area_ns <- data_area[-area_sampled$idx, ]
    m1 <- nrow(area_sampled)
    m2 <- nrow(area_ns)
    r <- c(area_sampled$idx, area_ns$idx)

    formuladata <- stats::model.frame(formula, data = unit_sampled, na.action = NULL)
    X_unit <- stats::model.matrix(formula, data = as.data.frame(formuladata))
    X_unit <- as.matrix(X_unit)

    formula_x <- stats::update(formula, NULL ~ .)
    X_area <- stats::model.matrix(formula_x, data = as.data.frame(area_sampled))
    X_area <- as.matrix(X_area)
    X_area_ns <- stats::model.matrix.lm(formula_x, data = as.data.frame(area_ns), na.action = stats::na.pass)
    X_area_ns <- as.matrix(X_area_ns)

    tau.ua <- tau.ub <- tau.ea <- tau.eb <- 1
    a.var <- 1

    my_model <- "model {
			for (j in 1:M1) {
					y[j] ~ dnorm(Mu[j], tau.e)
					Mu[j] <- b[1] + sum(b[2:nvar] * x[j,]) + u[d1[j]]
			}

			for(i in 1:m1){
			  u[i] ~ dnorm(0, tau.u)
			  mu[i] <- b[1] + sum(b[2:nvar] * X[i,]) + u[i]
			}

			for (k in 1:m2) {
				v[k] ~ dnorm(0, tau.u)
				muT[k] <- b[1] + sum(b[2:nvar] * XT[k,]) + v[k]
			}

			for (k in 1:nvar){
				 b[k] ~ dnorm(mu.b[k], tau.b[k])
			}

			tau.e ~ dgamma(tau.ea, tau.eb)
			tau.u ~ dgamma(tau.ua, tau.ub)
			a.var <- 1 / tau.u
	  }"

    for (iter in 1:iter.update) {
      update_progress(iter, iter.update)
      dat <- list(
        nvar = nvar,
        M1 = M1,
        m1 = m1, m2 = m2, d1 = d1,
        y = formuladata[, 1],
        x = as.matrix(X_unit[, -1]),
        X = as.matrix(X_area[, -1]),
        XT = as.matrix(X_area_ns[, -1]),
        mu.b = mu.b,
        tau.b = tau.b,
        tau.ua = tau.ua, tau.ub = tau.ub,
        tau.ea = tau.ea, tau.eb = tau.eb
      ) # names list of numbers

      inits <- list(
        u = rep(0, m1),
        v = rep(0, m2),
        b = mu.b,
        tau.u = 1,
        tau.e = 1
      )

      jags.m <- rjags::jags.model(
        file = textConnection(my_model),
        data = dat, inits = inits,
        n.chains = 1, n.adapt = 500, quiet = quiet
      )

      params <- c("mu", "muT", "a.var", "b", "tau.u", "tau.e")
      samps1 <- rjags::coda.samples(
        jags.m, params,
        n.iter = iter.mcmc, thin = thin,
        progress.bar = default_pb
      )
      # start is greater than n.adapt + burnin
      samps11 <- stats::window(samps1, start = burn.in + 1, end = iter.mcmc - 1)
      hasil <- summary(samps11)

      a.var <- hasil$statistics[1]
      beta <- hasil$statistics[2:(nvar + 1), 1:2]
      mu.b <- beta[, 1]
      tau.b <- 1 / (beta[, 2]^2)

      tau.e <- hasil$statistics[nvar + m + 2, 1:2]
      tau.u <- hasil$statistics[nvar + m + 3, 1:2]

      tau.ea <- tau.e[1]^2 / tau.e[2]^2
      tau.eb <- tau.e[1] / tau.e[2]^2
      tau.ua <- tau.u[1]^2 / tau.u[2]^2
      tau.ub <- tau.u[1] / tau.u[2]^2
    }


    result_samps <- summary(samps1)
    b.varnames <- c('intercept', all.vars(formula[[3]]))

    result_mcmc <- samps1[, c(2:(nvar + 1))]
    colnames(result_mcmc[[1]]) <- b.varnames
    a.var <- result_samps$statistics[1]
    beta <- result_samps$statistics[2:(nvar + 1), 1:2]
    rownames(beta) <- b.varnames
    mu <- result_samps$statistics[(nvar + 2):(1 + nvar + m), 1:2]

    Estimation <- data.frame(mu)
    Quantiles <- as.data.frame(result_samps$quantiles[2:(1 + nvar + m), ])
    q_mu <- Quantiles[-c(1:nvar), ]
    q_beta <- Quantiles[1:nvar, ]
    rownames(q_beta) <- b.varnames
    beta <- cbind(beta, q_beta)
    Estimation <- data.frame(Estimation, q_mu)
    colnames(Estimation) <- c(
      "MEAN", "SD", "2.5%", "25%",
      "50%", "75%", "97.5%"
    )
    # urutkan hasil estimasi seperti data awal
    Estimation <- dplyr::slice(Estimation, match(1:m, r))
    rownames(Estimation) <- .get_variable(data_area, domain)
  }

  result$Est <- Estimation
  result$coefficient <- beta
  result$refVar <- a.var
  result$result_mcmc <- result_mcmc
  class(result) <- 'saehb'

  if (plot) {
    graphics::par(mar = c(2, 2, 2, 2))
    coda::autocorr.plot(result_mcmc, col = "brown2", lwd = 2)
    plot(result_mcmc, col = "brown2", lwd = 2)
  }

  cli::cli_h1('Coefficient')
  stats::printCoefmat(beta)
  return(invisible(result))
}




# Fungsi bantuan ----------------------------------------------------------
.get_variable <- function(data, variable) {
  if (length(variable) == nrow(data)) {
    return(variable)
  } else if (methods::is(variable, "character")) {
    if (variable %in% colnames(data)) {
      variable <- data[[variable]]
    }else{
      cli::cli_abort('variable "{variable}" is not found in the data')
    }
  } else if (methods::is(variable, "formula")) {
    # extract column name (class character) from formula
    variable <- data[[all.vars(variable)]]
  } else {
    cli::cli_abort('variable "{variable}" is not found in the data')
  }
  return(variable)
}
